### ISSUE TITLE
Visibility of request details

### ISSUE THREAD

**Issue by jfock** [http-builder-ng/http-builder-ng#163]
Hi,

first of all, great job and thanks for the efforts in this cool tool! From my point of view there´s an issue with the visibility of request details. My intention is simply to log the details in a one-liner instead of using the client´s trace logging. In following example I´ve put a logger in the request config, in an interceptor and in a response handler:

```
        def client = ApacheHttpBuilder.configure {
            request.uri = 'http://this-service.net:12345'
            request.contentType = ContentTypes.JSON[0]
            request.accept = ContentTypes.JSON[0]
            request.headers['Authorization'] = 'Basic ' + "username:password".bytes.encodeBase64()
            request.charset = 'utf-8'

            execution.interceptor(GET) { cfg, fx ->
                LOGGER.info("interceptor: GET ${request.uri.toURI().toString()}${request.uri.path} ${request.headers}")
                fx.apply(cfg)
            }

            response.success {resp, reader ->
                LOGGER.info("response handler: GET ${request.uri.toURI().toString()}${request.uri.path} ${request.headers}")
                reader
            }
        }

        def response = client.get() {
            request.uri.path = "/api/path"
            request.headers['X-REQUEST-TOKEN'] = 'xyz'
            LOGGER.info("request config: GET ${request.uri.toURI().toString()}${request.uri.path} ${request.headers}")
        }
```

The output is as follows:
```
request config: GET http://this-service.net:12345/api/path [X-REQUEST-TOKEN:xyzj]
interceptor: GET http://this-service.net:12345 [Accept:application/json, Authorization:Basic PHduOm5zeW5j]
response handler: GET http://this-service.net:12345 [Accept:application/json, Authorization:Basic PHduOm5zeW5j]
```

Actually, these request headers were finally sent (captured by packet filter):
```
X-REQUEST-TOKEN: Test
Accept: application/json
Authorization: Basic PHduOm5zeW5j
Content-Type: application/json
Host: this-service.net:12345
Connection: Keep-Alive
User-Agent: Apache-HttpClient/4.5.2 (Java/1.8.0_112)
Accept-Encoding: gzip,deflate
```

Issues:
* I´ve set the contentType header, but it´s not visible in the http-builder request headers.
* I´ve set an additional header in the request config, but it´s not visible in the client´s request context. 
* And vice versa: the headers set in client config are not visible the request config.
* The URI path/query is not visible in the client´s request context
* I couldn't find out how to retrieve the request method 

Best working solution in this example is to put the logger into an interceptor. There I can use the ChainedHttpConfig to retrieve uri.path/query, but this config seems to hold only headers from the request config. So I would have to use both (HttpConfig and HttpChainedConfig) to fetch all request information 

```
execution.interceptor(GET) { cfg, fx ->
                LOGGER.info("interceptor: GET ${cfg.request.uri.toURI().toString()}${cfg.request.uri.path} ${cfg.request.headers} ${request.headers}")
                fx.apply(cfg)
            }
```

Feels all a bit inconsistent, because some details are visible (e.g. uri) others not, but maybe this is intended. Could you please check?!
Thanks!

**Comment by cjstehno** [comment]
Since each client implementation logs things a bit differently, we'd probably do something using an interceptor similar to what you are using to provide a common set of information across the clients. I will try and look at this soon - maybe I am missing something in your details. :-)

Note that only the client itself will have _all_ of the request information since HttpBuilder is just a configuration layer on top of other clients.

I am assuming you have read: https://http-builder-ng.github.io/http-builder-ng/asciidoc/html5/#_logging

**Comment by jfock** [comment]
My use case is an API with many different endpoints. Meaning, I have one more or less fix client configuration and a bunch of different request configs which may have additional headers, query parameters and even parser configuration when the API returns just a String instead JSON. And ideally I can place a logger somewhere in the response handler or interceptor to log the request I made. So my problem in short words is: I´m configuring a HTTP request and then I want to print/log the (complete) configuration which is at least difficult. 

For one HTTP request I have to deal with two configurations (client and request), the request settings are not visible in any case, e.g. the request method is known only in the request config, the Content-Type header is not added to the headers list (on http-builder level). Basically, settings I apply in one configuration are not (always) visible in the other. Obviously, the request configuration is merged at hand-over to the underlying client (which can be accessed only for customization), but not on http-builder level - maybe this could be an improvement.

And yes, the client´s debug logging provides the details I need, but according to the docs I´m also afraid that it dumps too much information to the logs and affects performance. But beside of the request logging my issue is pointing more on the http-builder configuration approach and the question if configuration should be transparent and visible (and in a next step maybe modifiable)

**Comment by cjstehno** [comment]
If we add a means of accessing the overall configuration at the various configuration levels, and perhaps a built-in means of logging it at request time, would this address your issue?

**Comment by jfock** [comment]
Yes, I think that would be really useful ;)

**Comment by cjstehno** [comment]
Development notes for myself:

* HttpBuilder needs a means of accessing its config (think its already there but be sure all accessible)
  * configure method should log current configuration at the DEBUG or TRACE level once applied
* Verb methods need to expose their configuration to the response handlers in some way
  * Verb methods should log current configuration at the DEBUG or TRACE level once applied
* Document this stuff in logging (along with the required logging categories to configure)

**Comment by cjstehno** [comment]
Ok, here is what I found (and did):

* The interceptors have access to the `ChainedHttpConfig` which has all the configuration (at all levels).
* I have added the request verb to that config so that it may be used in the logging.
* Interceptors are like servlet filters, they can do stuff on request and response (around the function apply call)
* I have added a bit more information to some of our internal logging

Due to the broad scope of the config we provide and its layered nature, I think it's best to keep this as something that is done in an interceptor as needed rather than in the main code.

The verb configuration methods are just that, configuration, the actual execution can be caught in an interceptor.

If this all still seems unclear, let me know and I could probably whip up a more detailed interceptor example. Also, let me know if you feel that I am missing one of your points.
